### PR TITLE
Thermal price definition (RTEi's -> CR20)

### PR DESCRIPTION
**Pull request by Milos-RTEi**
This PR is Identical to closed PR: https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1215
The transfer was done in order to use sonar-cloud.

===
This PR is created in order to implement RTEi development on CR20 (Thermal Price Definition).
Spec:
https://rteinternational.sharepoint.com/:w:/r/sites/AntaresUser%27sClub-Interne/Documents%20partages/Change%20Requests/Selected%20for%20development/CR20%20-%20Thermal%20Price%20Definition/Functional_Specifications-Thermal_Price_Definition_V0.1-HA.docx?d=w414939fbdb784b6f957f0bbc32e64a2d&csf=1&web=1&e=TYSOTg

### CODE REVIEW

File: src/libs/antares/study/parts/thermal/cluster.cpp [hunk 1]
```diff
@@ -421,6 +464,56 @@ void Data::ThermalCluster::calculationOfSpinning()
...
+        for (uint hour = 0; hour < HOURS_PER_YEAR; ++hour)
+        {
+            thermalEconomicTimeSeries[tsIndex].marketBidCostPerHourTs[hour]
+              = ecoInput->fuelcost[tsIndexFuel][hour] * 360.0 / fuelEfficiency
+                + emissions.factors[Pollutant::CO2] * ecoInput->co2cost[tsIndexCo2][hour]
+                + variableomcost;
+            thermalEconomicTimeSeries[tsIndex].marginalCostPerHourTs[hour]
+              = thermalEconomicTimeSeries[tsIndex].marketBidCostPerHourTs[hour];
+            if (modulation.width > 0)
+            {
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.cpp:436]: This should always return `true`, please remove this condition

File: src/libs/antares/study/parts/thermal/cluster.h [hunk 2]
```diff
@@ -346,6 +372,20 @@ class ThermalCluster final : public Cluster, public std::enable_shared_from_this
...
+    */
+    struct ThermalEconomicTimeSeries
+    {
+        std::array<double, HOURS_PER_YEAR> productionCostTs;
+        std::array<double, HOURS_PER_YEAR> marketBidCostPerHourTs;
+        std::array<double, HOURS_PER_YEAR> marginalCostPerHourTs;
+    };
+    std::vector<ThermalEconomicTimeSeries> thermalEconomicTimeSeries;
+
+    EconomicInputData* ecoInput;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.h]:
```suggestion
    EconomicInputData ecoInput;
```

File: src/ui/simulator/windows/inspector/frame.cpp [hunk 3]
```diff
@@ -525,13 +525,17 @@ Frame::Frame(wxWindow* parent, bool allowAnyObject) :
     pPGThClusterSpinning = P_FLOAT("Spinning (%)", "cluster.spinning");
     pPGThClusterSpinning->SetAttribute(wxPG_ATTR_MAX, 99.99);
     pPGThClusterCO2 = P_FLOAT("CO2 (Tons/MWh)", "cluster.co2");
-
+    pPGThClusterEfficiency = P_FLOAT("Efficiency (%)", "cluster.efficiency");
```
Comment (flomnes) [inline src/ui/simulator/windows/inspector/frame.cpp]:
```suggestion
    pPGThClusterEfficiency = P_FLOAT("Fuel efficiency (%)", "cluster.efficiency");
```

File: src/ui/simulator/windows/inspector/frame.cpp [hunk 4]
```diff
@@ -525,13 +525,17 @@ Frame::Frame(wxWindow* parent, bool allowAnyObject) :
...
+    pPGThClusterEfficiency->SetAttribute(wxPG_ATTR_MAX, 100.00);
+    
     pPGThClusterCosts = Category(pg, wxT("Operating costs"), wxT("cluster.costs"));
+    pPGThClusterCostGeneration = P_ENUM("Cost generation", "cluster.costgeneration", costgeneration);  
     pPGThClusterMarginalCost = P_FLOAT("Marginal (\u20AC/MWh)", "cluster.opcost_marginal");
     pPGThClusterFixedCost = P_FLOAT("Fixed (\u20AC/hour)", "cluster.opcost_fixed");
     pPGThClusterStartupCost = P_FLOAT("Startup (\u20AC/startup)", "cluster.opcost_startup");
     pPGThClusterOperatingCost = P_FLOAT("Market bid (\u20AC/MWh)", "cluster.opcost_marketbid");
     pPGThClusterRandomSpread = P_FLOAT("Spread (\u20AC/MWh)", "cluster.opcost_spread");
+    pPGThClusterVariableOMcost = P_FLOAT("Variable O\u0026M cost (\u20AC/MWh)", "cluster.variableomcost");
```
Comment (flomnes) [inline src/ui/simulator/windows/inspector/frame.cpp]:
```suggestion
    pPGThClusterVariableOMcost = P_FLOAT("Variable Operation\u0026Maintenance cost (\u20AC/MWh)", "cluster.variableomcost");
```

File: src/libs/antares/study/parts/thermal/cluster.cpp [hunk 5]
```diff
@@ -421,6 +464,56 @@ void Data::ThermalCluster::calculationOfSpinning()
     }
 }
 
+void Data::ThermalCluster::calculationOfMarketBidPerHourAndMarginalCostPerHour()
+{
+    if (costgeneration == Data::setManually || !prepro)
+        costGenManualCalculationOfMarketBidAndMarginalCostPerHour();
+    else // costgeneration == Data::useCostTimeseries
+        costGenTimeSeriesCalculationOfMarketBidAndMarginalCostPerHour();
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.cpp]:
```suggestion
    switch (costgeneration) {
    case Data::setManually:
        costGenManualCalculationOfMarketBidAndMarginalCostPerHour();
        break;
    case Data::useCostTimeseries:
        costGenTimeSeriesCalculationOfMarketBidAndMarginalCostPerHour();
        break;
```
Note that `prepro` is now completely unrelated.

File: docs/reference-guide/04-active_windows.md [hunk 6]
```diff
@@ -248,6 +248,8 @@ Durations are expressed in days and rates belong to [0 , 1].
     - Minimum Down time (hours)
     - Default contribution to the spinning reserve (% of nominal capacity)
     - CO2 tons emitted per electric MWh
+    - Efficiency (%)
```
Comment (flomnes) [inline docs/reference-guide/04-active_windows.md]:
```suggestion
    - Fuel efficiency (%)
```

File: docs/reference-guide/04-active_windows.md [hunk 7]
```diff
@@ -261,6 +263,7 @@ Durations are expressed in days and rates belong to [0 , 1].
     - Start-up cost (€/start-up)
     - Market bid (€/MWh)
     - Random spread on the market bid (€/MWh)
+    - Variable O&M cost (€/MWh)
```
Comment (flomnes) [inline docs/reference-guide/04-active_windows.md]:
```suggestion
    - Variable Operation&Maintenance cost (€/MWh, only used if **Cost generation** is set to **use cost timeseries**)
```

File: docs/reference-guide/13-file-format.md [hunk 8]
```diff
@@ -26,6 +26,14 @@ A few changes related to the introduction of short-term storage objects.
 
 In files **input/thermal/cluster/area/list.ini** add properties `nh3`, `nox`, `pm2_5`, `pm5`, `pm10`, `nmvoc`, `op1`, `op2`, `op3`, `op4`, `op5` [double]. These properties are emission factors similar to the existing one for CO2.
 
+For each thermal cluster, in existing file **input/thermal/clusters/&lt;area&gt;/list.ini**, under existing sections **&lt;cluster&gt;**, following properties added: 
+
+* `costgeneration` [string] can take values `useCostTimeseries` or be excluded from the section if `Set manually` is selected (default behavior).
+* `efficiency` [float] excluded from the section if default value 100 is selected (default behavior).
+* `variableomcost` [float] excluded from the section if default value 0 is selected (default behavior).
+
+For each thermal cluster, new files added **input/thermal/prepro/&lt;area&gt;/&lt;cluster&gt;/CO2Cost.txt** and **input/thermal/series/&lt;area&gt;/&lt;cluster&gt;/fuelCost.txt**.
```
Comment (flomnes) [inline docs/reference-guide/13-file-format.md]:
```suggestion
For each thermal cluster, new files added **input/thermal/prepro/&lt;area&gt;/&lt;cluster&gt;/CO2Cost.txt** and **input/thermal/series/&lt;area&gt;/&lt;cluster&gt;/fuelCost.txt**. **fuelCost.txt** and **CO2Cost.txt** must either have one column, or the same number of columns as existing file **series.txt** (availability)
```

File: src/libs/antares/study/parts/thermal/cluster_list.cpp [hunk 9]
```diff
@@ -355,6 +376,8 @@ void ThermalClusterList::ensureDataPrepro()
         auto c = it->second;
         if (not c->prepro)
             c->prepro = new PreproThermal(c);
+        if (!c->ecoInput)
+            c->ecoInput = new EconomicInputData(c);
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster_list.cpp]: If the pointer is removed, this is no longer necessary

File: src/libs/antares/study/parts/thermal/cluster.cpp [hunk 10]
```diff
@@ -180,6 +208,7 @@ Data::ThermalCluster::~ThermalCluster()
 {
     delete[] productionCost;
     delete prepro;
+    delete ecoInput;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.cpp]: If the pointer is removed, this is no longer necessary

File: src/libs/antares/study/parts/thermal/cluster.cpp [hunk 11]
```diff
@@ -421,6 +464,56 @@ void Data::ThermalCluster::calculationOfSpinning()
...
+
+void Data::ThermalCluster::costGenManualCalculationOfMarketBidAndMarginalCostPerHour()
+{
+    std::fill(thermalEconomicTimeSeries[0].marketBidCostPerHourTs.begin(),
+              thermalEconomicTimeSeries[0].marketBidCostPerHourTs.end(),
+              marketBidCost);
+    std::fill(thermalEconomicTimeSeries[0].marginalCostPerHourTs.begin(),
+              thermalEconomicTimeSeries[0].marginalCostPerHourTs.end(),
+              marginalCost);
+    return;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.cpp]:
```suggestion
```

File: src/libs/antares/study/parts/thermal/cluster.cpp [hunk 12]
```diff
@@ -421,6 +464,56 @@ void Data::ThermalCluster::calculationOfSpinning()
...
+            if (modulation.width > 0)
+            {
+                thermalEconomicTimeSeries[tsIndex].productionCostTs[hour]
+                  = thermalEconomicTimeSeries[tsIndex].marginalCostPerHourTs[hour]
+                    * modulation[Data::thermalModulationCost][hour];
+            }
+        }
+    }
+
+    return;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.cpp]:
```suggestion
```

File: src/libs/antares/study/parts/thermal/cluster.cpp [hunk 13]
```diff
@@ -421,6 +464,56 @@ void Data::ThermalCluster::calculationOfSpinning()
...
+        uint tsIndexFuel = std::min(fuelCostWidth - 1, tsIndex);
+        uint tsIndexCo2 = std::min(co2CostWidth - 1, tsIndex);
+        for (uint hour = 0; hour < HOURS_PER_YEAR; ++hour)
+        {
+            thermalEconomicTimeSeries[tsIndex].marketBidCostPerHourTs[hour]
+              = ecoInput->fuelcost[tsIndexFuel][hour] * 360.0 / fuelEfficiency
+                + emissions.factors[Pollutant::CO2] * ecoInput->co2cost[tsIndexCo2][hour]
+                + variableomcost;
+            thermalEconomicTimeSeries[tsIndex].marginalCostPerHourTs[hour]
+              = thermalEconomicTimeSeries[tsIndex].marketBidCostPerHourTs[hour];
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster.cpp]: Please isolate this computation in a function such as `computeMarketBidCost(double fuelCost, double co2EmissionFactor, double co2cost, double variableomcost)` to increase readability.

File: src/ui/simulator/windows/inspector/frame.cpp (hunk 3 above)
Comment (flomnes) [reply src/ui/simulator/windows/inspector/frame.cpp]: Also, please move this key/value into section **Operating costs**

image

**Review by flomnes — COMMENTED** [review]
The thermal costs are getting a bit complicated. Here are the inputs
- CO2 (Tons/MWh)
- Marginal cost (€/MWh)
- Fixed cost (€/MWh)
- Startup cost (€/MWh)
- Market bid (€/MWh)
- Spread (€/MWh)
- Marginal cost modulation
- Market bid modulation
- Variable O&M cost (€/MWh) {NEW}
- Fuel cost (€/GJ) {NEW}
- CO2 cost (€/ton) {NEW}

We need to make it clear what data is used for the optimization, and what data is used to compute the costs in post-processing.

This complexity is getting out of hand IMO, although you're not to blame. This part needs more documentation.

File: src/libs/antares/study/parts/thermal/cluster.cpp (hunk 1 above)
Comment (Milos-RTEi) [reply src/libs/antares/study/parts/thermal/cluster.cpp:436]: image
Failing to load the study if removed!

File: src/libs/antares/study/parts/thermal/cluster_list.cpp [hunk 14]
```diff
@@ -139,8 +143,19 @@ bool ThermalClusterList::loadFromFolder(Study& study, const AnyString& folder, A
...
+                // alias to the marginal cost
+                for (uint h = 0; h != cluster->modulation.height; ++h)
+                    prodCost[h] = marginalCost * modulation[h];
+            }
+            else
+            {
+                const auto& marginalCostPerHour
+                  = cluster->thermalEconomicTimeSeries[0].marginalCostPerHourTs;
+                for (uint h = 0; h != cluster->modulation.height; ++h)
+                    prodCost[h] = marginalCostPerHour[h] * modulation[h];
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/cluster_list.cpp:157]: You use modulation time-series only for `prodCost` if `costgeneration == useTimeSeries`. `modulation[thermalModulationMarketBid]` is ignored, is this intentional ?

File: src/libs/antares/study/parts/thermal/ecoInput.cpp [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+
+    return ret;
+}
+
+bool EconomicInputData::forceReload(bool reload) const
+{
+    bool ret = true;
+    ret = fuelcost.forceReload(reload) && ret;
+    ret = co2cost.forceReload(reload) && ret;
+    return ret;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/ecoInput.cpp]:
```suggestion
    return fuelcost.forceReload(reload) && co2cost.forceReload(reload);
```

File: src/libs/antares/study/parts/thermal/ecoInput.cpp [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+bool EconomicInputData::loadFromFolder(Study& study, const AnyString& folder)
+{
+    bool ret = true;
+    auto& buffer = study.bufferLoadingTS;
+
+    auto cluster = itsThermalCluster.lock();
+    if (!cluster)
+        return false;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/ecoInput.cpp]: I can't see why `cluster` is needed here. `itsThermalCluster` doesn't seem to be read elsewhere, so I wonder if it's really needed as well.

File: src/libs/antares/study/parts/thermal/cluster_list.cpp (hunk 14 above)
Comment (flomnes) [reply src/libs/antares/study/parts/thermal/cluster_list.cpp:157]: See **sim_calcul_economique.cpp**

**Comment by flomnes** [comment]
We removed `unitCountLastHour`, `productionLastHour` and `pminOfAGroup` and a constructor of `Data::ThermalCluster`

File: src/libs/antares/study/parts/thermal/prepro.cpp [hunk 17]
```diff
@@ -168,7 +167,9 @@ bool PreproThermal::loadFromFolder(Study& study, const AnyString& folder)
 
 bool PreproThermal::forceReload(bool reload) const
 {
-    return data.forceReload(reload);
+    bool ret = true;
+    ret = data.forceReload(reload) && ret;
+    return ret;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/prepro.cpp]:
```suggestion
    return data.forceReload(reload);
```

File: src/libs/antares/study/parts/thermal/ecoInput.hxx [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+namespace Antares::Data
+{
+inline Yuni::uint64 EconomicInputData::memoryUsage() const
+{
+    return sizeof(EconomicInputData);
+}
+
+} // namespace Antares::Data
+
+#endif // __ANTARES_LIBS_STUDY_PARTS_THERMAL_ECOINPUT_HXX__
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/ecoInput.hxx]: Move to .cpp file. The idea is to place only templated functions inside hxx files.

File: src/libs/antares/study/parts/thermal/prepro.cpp [hunk 19]
```diff
@@ -57,11 +57,13 @@ void PreproThermal::copyFrom(const PreproThermal& rhs)
 
 bool PreproThermal::saveToFolder(const AnyString& folder)
 {
+    bool ret = true;
```
Comment (flomnes) [inline src/libs/antares/study/parts/thermal/prepro.cpp]: You don't need `ret`

**Comment by flomnes** [comment]
Please fix conflicts

**Review by flomnes — APPROVED** [review]

**Pull request merged**